theano/gpuarray/dnn.py

```diff
@@ -2738,6 +2738,342 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+
+
+class GpuDnnTransformerSampler(DnnBase):
+    """
+    Grid sampler Op for cuDNN Spatial Transformer.
+    """
+    __props__ = ()
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+    _f16_ok = True
+    check_input = False
+
+    def __init__(self):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_sampler.c"], "APPLY_SPECIFIC(dnn_sptf_sampler)")
+
+    def make_node(self, img, grid, desc):
+        """
+        Create a grid sampler node for a cuDNN Spatial Transformer
+
+        Parameters
+        ----------
+        img : tensor
+            Images from which the pixels will be sampled. The implementation
+            assumes the tensor is in NCHW format, where N is the number of images,
+            C is the number of color channels, H is the height of the inputs, and
+            W is width of the inputs.
+
+        grid : GpuDnnTransformerGrid
+            Grid that contains the coordinates of the pixels to be sampled from
+            the inputs images.
+
+        desc : GpuDnnTransformerDesc
+            Spatial transformer descriptor
+        """
+        context_name = infer_context_name(desc)
+
+        if (not isinstance(desc.type, CDataType) or
+                desc.type.ctype != 'cudnnSpatialTransformerDescriptor_t'):
+            raise ValueError('desc must be cudnnSpatialTransformerDescriptor_t')
+
+        img = gpu_contiguous(as_gpuarray_variable(img, context_name))
+        if img.type.ndim != 4:
+            raise TypeError('img must be a 4D tensor')
+        elif img.dtype not in ('float16', 'float32', 'float64'):
+            raise TypeError('img type must be floating-point')
+
+        grid = gpu_contiguous(as_gpuarray_variable(grid, context_name))
+        if grid.type.ndim != 4:
+            raise TypeError('grid must be a 4D tensor')
+        elif grid.dtype not in ('float16', 'float32', 'float64'):
+            raise TypeError('grid type must be floating-point')
+
+        out = GpuArrayType(dtype=img.dtype,
+                           broadcastable=img.type.ndim * (False,),
+                           context_name=context_name)()
+
+        inputs = [img, grid, desc]
+        outputs = [out]
+        return Apply(self, inputs, outputs)
+
+    def grad(self, inputs, grads):
+        img, grid, desc = inputs
+        dy = grads[0]
+
+        dimg, dgrid = GpuDnnTransformerGradI()(img, grid, dy, desc)
+        return [dimg, dgrid, DisconnectedType()()]
+
+    def connection_pattern(self, node):
+        # not connected to desc
+        return [[1], [1], [0]]
+
+
+class GpuDnnTransformerGradI(DnnBase):
+    """
+    Gradient of inputs Op for cuDNN Spatial Transformer.
+    """
+    __props__ = ()
+    _cop_num_inputs = 4
+    _cop_num_outputs = 2
+    _f16_ok = True
+    check_input = False
+
+    def __init__(self):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gi.c"], "APPLY_SPECIFIC(dnn_sptf_gi)")
+        if theano.config.floatX == 'float16' or theano.config.floatX == 'float32':
```

Comment:
I've removed the warnings, there is just a problem with the gradients on float16 not being under the tolerance (order of 1 or 10).